scripts/30_prepare_disk.sh

```diff
@@ -208,6 +208,51 @@ partition_resize()
     sfdisk --quiet "${TARGET_STORAGE_DEVICE}" < "${SYSTEM_UPDATE_DIR}/${PARTITION_TABLE_FILE}"
 }
 
+backup_data()
+{
+    temp_mount_dir=$(mktemp -d)
+
+    sfdisk --quiet --dump "${TARGET_STORAGE_DEVICE}" | \
+    while IFS="${IFS}:=," read -r disk_device _ disk_start _ disk_size _; do
+        if grep -q "${disk_device}" /proc/mounts; then
+            umount "${disk_device}"
+        fi
+        backup_file="/tmp/backup${disk_device}.tar.gz"
+        if mount "${disk_device}" "${temp_mount_dir}"; then
+            mkdir -p "$(dirname "${backup_location}")"
+            if ! tar -czf "${backup_location}" -C "${temp_mount_dir}" .; then
+                rm "${backup_location}"
+            fi
+            umount "${disk_device}"
+        fi
+    done
+
+    rmdir "${temp_mount_dir}"
+}
+
+restore_data()
+{
+    temp_mount_dir=$(mktemp -d)
+
+    sfdisk --quiet --dump "${TARGET_STORAGE_DEVICE}" | \
+    while IFS="${IFS}:=," read -r disk_device _ disk_start _ disk_size _; do
+        if grep -q "${disk_device}" /proc/mounts; then
+            umount "${disk_device}"
+        fi
+        backup_file="/tmp/backup${disk_device}.tar.gz"
+        if [ -f "${backup_file}" ]; then
+            mount "${disk_device}" "${temp_mount_dir}"
+            # We only restore if the parition looks empty. There can be a lost+found on an empty partition, so ignore that.
+            if [ "$(ls "${temp_mount_dir}" | wc -l)" < 2; then
```

Comment:
Shellcheck SC2012: Use find instead of ls to better handle non-alphanumeric filenames.